components/raftstore/src/store/fsm/store.rs

```diff
@@ -1766,13 +1781,29 @@ impl<EK: KvEngine, ER: RaftEngine> RaftBatchSystem<EK, ER> {
         let region_scheduler = workers
             .region_worker
             .start_with_timer("region-worker", region_runner);
+        // Same as split_check_scheduler, region worker also runs a infinite
+        // loop, that will not stop when shutting down the threadpool, causing the
+        // reference to the kv engine inside the scheduler to be dangle. So we need
+        // to stop the scheduler before shutting down the threadpool.
+        let region_scheduler_clone = region_scheduler.clone();
```

Comment:
FYI, `snap_gen_worker` also owns the engine reference, should we stop the scheduler in advance before shutting down the relevant workers?

Reply:
`snap_gen_worker` is not self-referencing, once all the references are dropped, it stops. The 3 scheduelrs / workers are self-referencing, they have a reference to the channel sender, causing the channel never stop.

Reply:
@v01dstar Could you please explain how to determine if a structure is self-referencing? Both `region_worker` and `snap_gen_worker` are `Worker`, it seems there is no difference except for its generic args.

Reply:
They have channel senders whose receivers are themselves

Reply:
Is the channel used in `start_with_timer_impl`?

https://github.com/tikv/tikv/blob/f02d2904696ff6276ccc6f86dbeb3a557df3b9ce/components/tikv_util/src/worker/pool.rs#L525

Reply:
Yes